### PR TITLE
add generic kernel cache

### PR DESCRIPTION
**Pull request by gartnera**
Add a generic version of #1577. Should be pretty easy to expand to metal too. The `compile` method must return the bytes of the program binary. `compile_cache` decorator returns a path to that binary.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/runtime/lib.py                   38               1
tinygrad/runtime/ops_clang.py             36              25
tinygrad/runtime/ops_hip.py               16              10
------------------------------------------------------------
total                                     90              36
------------------------------------------------------------
lines added in the tinygrad folder: 54
```

File: tinygrad/runtime/lib.py [hunk 1]
```diff
@@ -109,3 +113,36 @@ def _underlying_buf_memsz(self, buf): return self.buffer_info[buf][0] * self.buf
   def _cached_bufkey(self, size, dtype, device) -> Tuple[int, ...]: return (device, size, dtype, dtype.shape) if isinstance(dtype, ImageDType) else (device, size, dtype) # Provides a key for reusing device buffers with identical keys.
   def _do_alloc(self, size, dtype, device, **kwargs): raise NotImplementedError("must be implemented")
   def _do_free(self, buf): pass
+
+class CachedProgram(abc.ABC):
```
Comment (geohot) [inline tinygrad/runtime/lib.py:117]: Err, not like this. Make a decorator for the compile function, don't extend the Program class

**Comment by geohot** [comment]
Tests for the cache? And is it possible to do this with a negative total line diff?

**Comment by gartnera** [comment]
> Tests for the cache?

will do

> And is it possible to do this with a negative total line diff?

I don't see a way. Could slightly reduce by removing the toolchain hash bit but that's pretty important for cache coherency.

File: tinygrad/runtime/ops_hip.py [hunk 2]
```diff
@@ -14,6 +14,9 @@
 
 # The default HIP stream is used for everything.
 
+def toolchain_hash():
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:17]: example, this should be one line and on line 41

File: tinygrad/runtime/lib.py [hunk 3]
```diff
@@ -109,3 +113,29 @@ def _underlying_buf_memsz(self, buf): return self.buffer_info[buf][0] * self.buf
...
+  cache_dir = os.path.join(cache_dir_base, f"{prefix}-{toolchain_hash}")
+  def compile_cache_1(func):
+    def compile_cache_2(self, name:str, prg:str, binary:bool=False, extension:str="") -> str:
+      prg_hash = hashlib.sha256(prg.encode()).digest().hex()
+      bin_cache_path = os.path.join(cache_dir, prg_hash)
+      if extension:
+        bin_cache_path = f"{bin_cache_path}.{extension}"
+      if not os.path.exists(bin_cache_path):
+        bin_cache_path_tmp = bin_cache_path + f".tmp.{os.getpid()}"
+        os.makedirs(cache_dir, exist_ok=True)
```
Comment (geohot) [inline tinygrad/runtime/lib.py:128]: This should not be called every time

File: tinygrad/runtime/lib.py [hunk 4]
```diff
@@ -109,3 +113,29 @@ def _underlying_buf_memsz(self, buf): return self.buffer_info[buf][0] * self.buf
...
   def _do_free(self, buf): pass
+
+def compile_cache(prefix:str, toolchain_hash:str):
+  cache_dir_base = os.path.expanduser("~/.cache/tinygrad-programs")
+  cache_dir = os.path.join(cache_dir_base, f"{prefix}-{toolchain_hash}")
+  def compile_cache_1(func):
+    def compile_cache_2(self, name:str, prg:str, binary:bool=False, extension:str="") -> str:
+      prg_hash = hashlib.sha256(prg.encode()).digest().hex()
+      bin_cache_path = os.path.join(cache_dir, prg_hash)
+      if extension:
```
Comment (geohot) [inline tinygrad/runtime/lib.py:124]: This if statement is useless

File: tinygrad/runtime/ops_clang.py [hunk 5]
```diff
@@ -29,31 +35,41 @@ def emulate_ext_calls(fn, uc, address, size, user_data):
...
-        os.rename(tmp, fn)
-      else:
+    if not binary:
+      prg = CLANG_PROGRAM_HEADER + prg
+      with mktemp_clean() as path:
+        subprocess.check_output(args=('clang -shared -O2 -Wall -Werror -x c '+args['cflags']+' - -o '+path).split(), input=prg.encode('utf-8'))
+        with open(path, "rb") as f:
+          return f.read()
+    else:
+      with mktemp_clean() as as_path, mktemp_clean() as final_path:
```
Comment (geohot) [inline tinygrad/runtime/ops_clang.py:62]: Is there a way to pipe this?

Comment (gartnera) [reply tinygrad/runtime/ops_clang.py:62]: With clang I tried to write it to `/dev/stdout` but you get `/usr/bin/ld: final link failed: Illegal seek`.

Oddly enough `cat test.c | clang -shared -O2 -Wall -Werror -x c -lm -fPIC --rtlib=compiler-rt - -o /dev/stdout > test.so` seems to work.

Something like `res = subprocess.check_output(['dd', 'if=/dev/stdin', 'of=/dev/stdout'], input=prg)` also works.

**Comment by wozeparrot** [comment]
Any update on this?

**Comment by wozeparrot** [comment]
Closing for inactivity.

**Pull request closed without merging**